src/Sarif/Readers/PropertyBagConverter.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.CodeAnalysis.Sarif.Readers
+{
+    /// <summary>
+    /// Converts a property bag (a JSON object whose keys have arbitrary names and whose values
+    /// may be any JSON values) into a dictionary whose keys match the JSON object's
+    /// property names, and whose values are of type <see cref="SerializedPropertyInfo"/>
+    /// </summary>
+    internal class PropertyBagConverter : JsonConverter
+    {
+        internal static readonly JsonConverter Instance = new PropertyBagConverter();
+
+        public override bool CanConvert(Type objectType)
+        {
+            return typeof(IDictionary<string, SerializedPropertyInfo>).IsAssignableFrom(objectType);
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            var objectDictionary = new Dictionary<string, object>();
+            serializer.Populate(reader, objectDictionary);
+
+            var propertyDictionary = new Dictionary<string, SerializedPropertyInfo>();
+            foreach (string key in objectDictionary.Keys)
+            {
+                JTokenType jTokenType = JTokenType.Undefined;
+                Type propertyType = objectDictionary[key].GetType();
+
+                if (propertyType == typeof(JObject))
+                {
+                    jTokenType = JTokenType.Object;
+                }
+                else if (propertyType == typeof(JArray))
+                {
+                    jTokenType = JTokenType.Array;
+                }
+                else if (propertyType == typeof(bool))
+                {
+                    jTokenType = JTokenType.Boolean;
+                }
+                else if (propertyType == typeof(long))
+                {
+                    jTokenType = JTokenType.Integer;
+                }
+                else if (propertyType == typeof(double))
+                {
+                    jTokenType = JTokenType.Float;
+                }
+                else if (propertyType == typeof(string))
+                {
+                    jTokenType = JTokenType.String;
+                }
```

Comment:
I need to think about that. JSON doesn't define (for example) a Guid type, so in the file you'd just have a string:

{
  "properties": {
    "g": "{12345678-90ab-cdef-1234-567890abcdef}"
  }
}

That means that when the converter gets hold of it, all it will know is that there's a string. You can ask for it as a Guid and that will work. I just wrote this unit test and it passes:

            _input = "{\"properties\":{\"g\":\"{12345678-90ab-cdef-1234-567890abcdef}\"}}";

            PerformRoundTrip();

            Guid expectedGuid = new Guid("{12345678-90ab-cdef-1234-567890abcdef}");

            _inputObject.GetProperty<Guid>("g").Should().Be(expectedGuid);

The same should work for Uris and DateTimes; I'll check.

---
In reply to: [62770570](https://github.com/Microsoft/sarif-sdk/pull/257#discussion_r62770570) [](ancestors = 62770570)